docs/source/cuda/cuda_array_interface.rst

```diff
@@ -69,13 +69,249 @@ The following are optional entries:
+    provided stream, then it may return control to the User code immediately
+    after enqueueing its work, as the work will all be serialized on the
+    exported array's stream. This is sufficient to ensure correctness even if
+    the User code were to induce the Producer to subsequently start enqueueing
+    more work on the same stream.
+
+* If the User has set the Consumer to ignore CAI synchronization semantics, the
+  Consumer may assume it can operate on the data immediately in any stream with
+  no further synchronization, even if the ``stream`` member has an integer
+  value.
+
+
+When exporting an array through the CAI, Producers must ensure that:
+
+* If there is work on the data enqueued in one or more streams, then
+  synchronization on the provided ``stream`` is sufficient to ensure
+  synchronization with all pending work.
+
+  - If the Producer has no enqueued work, or work only enqueued on the stream
+    identified by ``stream``, then this condition is met.
+  - If the Producer has enqueued work on the data on multiple streams, then it
+    must enqueue events on those streams that follow the enqueued work, and
+    then wait on those events in the provided ``stream``. For example:
+
+    1. Work is enqueued by the Producer on streams ``7``, ``9``, and ``15``.
+    2. Events are then enqueued on each of streams ``7``, ``9``, and ``15``.
+    3. Waits on the events from Step 2 are enqueued into stream ``3``, and the
+       ``stream`` entry is set to ``3``.
+
+* If there is no work enqueued on the data, then the ``stream`` entry may be
+  either ``None``, or not provided.
+
+Optionally, to facilitate the User to relax the conformance to synchronization
+semantics:
+
+* Producers may provide a configuration option to always set ``stream`` to
+  ``None``.
+* Consumers may provide a configuration option to ignore the value of ``stream``
+  and act as if it were ``None`` or not provided.  This elides synchronization
+  on the Producer-provided streams, and allows enqueuing work on streams other
+  than that provided by the Producer.
+
+These options should be not be set by default in either a Producer or a
+Consumer. The exact mechanism by which these options are set, and related
+options that Producers or Consumers might provide to allow the user further
+control over synchronization behavior are not prescribed by the CAI
+specification.
+
+
+Synchronization in Numba
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Numba is neither strictly a Producer nor a Consumer - it may be used to
+implement either by a User. In order to facilitate the correct implementation of
+synchronization semantics, Numba exhibits the following behaviors related to
+synchronization of the interface:
+
+- When Numba acts as a Consumer (for example when an array-like object is passed
+  to a kernel launch): If ``stream`` is an integer, then Numba will immediately
+  synchronize on the provided ``stream``, and set the *default stream* of the
+  array to the provided stream.
```

Comment:
Why does Numba need to synchronize the stream? It could just launch the kernel on the same stream and then require no synchronization, no?

Reply:
A kernel can be called with multiple arguments, that might have different streams, e.g.:

```python
arr1.__cuda_array_interface__
# {'shape': (10,),
#  'strides': None,
#  'data': (140217113640960, False),
#  'typestr': '<f8',
#  'stream': 94913095065792,
#  'version': 3}

arr2.__cuda_array_interface__
# {'shape': (10,),
#  'strides': None,
#  'data': (140217113641472, False),
#  'typestr': '<f8',
#  'stream': 94913095181664,
#  'version': 3}

kernel[griddim, blockdim](arr1, arr2)
```

Also, the user might want to launch the kernel on a different stream, e.g. with:

```python
# (Note that Numba configurations are griddim, blockdim, stream, dynsharedmemsize, unlike CUDA C/C++)
kernel[griddim, blockdim, 0, stream](arr1, arr2)
```

Reply:
Thanks @gmarkall , that makes total sense, I didn't consider more complex cases such as the example you wrote.

Reply:
Is it Numba's responsibility to enforce this for the user? I.E. if read an array with a stream using `__cuda_array_interface__` and then launch a kernel on a different stream, is it Numba's responsibility to make that safe?

From the perspective of a library using Numba internally, I'd like to be able to just stream order everything and avoid synchronizing.

Reply:
It is, because of the requirement that interactions between Producers and Consumers should be correct by default (rather than relying on the User / User Code to do the right thing). I think the way to go here is to add a config flag to Numba that controls whether Numba enforces this or not, so that a naive use of the interface (calling a kernel on an object which supplies a stream) synchronizes and ensures correctness, but users who want more control can avoid the sync and just stream order everything. I'll add a flag for this to this PR, and also the code that syncs on kernel launch dependent on the value of the flag (which is presently missing from the implementation in this PR).

Reply:
Alternatively, how hard would it be to have Numba scan the `stream` of all input array(s) and sychronize those stream(s) before the kernel launch? Or we could insert events into those stream(s) and then wait on those events in the stream of the kernel launch. If the stream(s) of the input array(s) is the same as the stream of the kernel launch then there's no need to synchronize or wait on events. This would maintain safety while also removing unneeded synchronization.

Reply:
It would be possible to do that, but it would be a bit of work to get the right information into the right places in Numba. For the purpose of this PR, I'd prefer to focus on:

1. Getting the specification right, for all libraries that use it
2. Coming up with a basic implementation in Numba that meets the requirements

then once the specification and basic Numba implementation are merged, then working on optimizing the Numba implementation.

To comment a bit more on the suggestion - I think in addition to synchronizing on the streams (or waiting on events on the array streams in the kernel stream), we'd also need to ensure synchronization following the kernel, because it launches asynchronously - so we'd have to insert an event in the kernel's stream after its launch, and waits in the streams of any of the arrays accessed via the CUDA Array Interface.

Reply:
> From the perspective of a library using Numba internally, I'd like to be able to just stream order everything and avoid synchronizing.

This should now be doable in this branch by setting `numba.config.CUDA_ARRAY_INTERFACE_SYNC = 0`.

Reply:
Having the `sync` flag in `as_cuda_array` / `from_cuda_array_interface` makes this less of a concern.

If someone passes a non-numba object with CAI into a kernel call I guess it would sync for now unless the config is explicitly set?

Reply:
Yes, that would happen now unless the config flag is set.

Reply:
... I'm trying to think if there is a non-horrible way to allow the option of a `sync=` flag to a kernel launch itself though.

Reply:
@kkraus14 for a non-syncing launch, it looks like it would be doable to replace the spelling:

```python
kernel[griddim, blockdim, stream, sharedmem](*args)
```

with

```python
kernel.configure(griddim, blockdim, stream=stream, sharedmem=sharedmem, sync=False)(*args)
```

(noting that `stream` and `sharedmem` are optional) - this would give some more granular control over whether a kernel launch syncs on the streams of arrays passed in through the CAI.

Would this be likely to be worth the effort to support a use case where syncs are disabled for individual launches, as opposed to globally?

Reply:
My view is that having `numba.as_cuda_array(obj, sync=False)` gives an easy enough / clear enough path to avoid synchronizing the stream(s) before kernel launch. It wouldn't synchronize if passed numba device arrays, correct?

Reply:
That's correct, it won't synchronize if passed Numba device arrays.

Reply:
Okay sounds great. I don't think it makes sense to put in the effort to add a `sync` kwarg to the kernel configuration then for now and we can revisit once we decide on the synchronization behavior we want to enforce more generally in Numba.